cpp/src/datetime/datetime_ops.cu

```diff
@@ -127,6 +127,25 @@ struct extract_day_num_of_year {
   }
 };
 
+// Extract the day number of the year present in the timestamp
+struct extract_quarter_op {
+  template <typename Timestamp>
+  CUDA_DEVICE_CALLABLE int16_t operator()(Timestamp const ts) const
+  {
+    using namespace cuda::std::chrono;
+
+    // Only has the days - time component is chopped off, which is what we want
+    auto const days_since_epoch = floor<days>(ts);
+    auto const date             = year_month_day(days_since_epoch);
+    auto const month            = unsigned{date.month()};
+
+    if (month <= 3) return 1;
+    if (month <= 6) return 2;
+    if (month <= 9) return 3;
+    return 4;
```

Comment:
Maybe it just my preference but using an `int16_t` to represent a [sum type](https://en.wikipedia.org/wiki/Tagged_union) with only four options (Q1, Q2, Q3 and Q4) seems suboptimal. I would recommend an `enum class`.

Reply:
Instead of using multiple if conditions, maybe just `round(month/3)` or similar?

Reply:
Imo, `int16_t` makes sense here as it makes the work for the python layer a lot easier and also it allows the rounding trick mentioned by @ttnghia.